### PR TITLE
Add torch.autograd.Function

### PR DESCRIPTION
**Pull request by shaltielshmid**

### CODE REVIEW

**Comment by shaltielshmid** [comment]
Linking to issue: https://github.com/dotnet/TorchSharp/issues/606

**Comment by shaltielshmid** [comment]
Example code in C++ (taken from https://pytorch.org/tutorials/advanced/cpp_autograd.html):

```cpp
class LinearFunction : public Function<LinearFunction> {
 public:
  // Note that both forward and backward are static functions

  // bias is an optional argument
  static torch::Tensor forward(
      AutogradContext *ctx, torch::Tensor input, torch::Tensor weight, torch::Tensor bias = torch::Tensor()) {
    ctx->save_for_backward({input, weight, bias});
    auto output = input.mm(weight.t());
    if (bias.defined()) {
      output += bias.unsqueeze(0).expand_as(output);
    }
    return output;
  }

  static tensor_list backward(AutogradContext *ctx, tensor_list grad_outputs) {
    auto saved = ctx->get_saved_variables();
    auto input = saved[0];
    auto weight = saved[1];
    auto bias = saved[2];

    auto grad_output = grad_outputs[0];
    auto grad_input = grad_output.mm(weight);
    auto grad_weight = grad_output.t().mm(input);
    auto grad_bias = torch::Tensor();
    if (bias.defined()) {
      grad_bias = grad_output.sum(0);
    }

    return {grad_input, grad_weight, grad_bias};
  }
};
```

Equivalent in C#:

```cs
class LinearFunction : torch.autograd.SingleTensorFunction<LinearFunction>
{
    public override string Name => nameof(LinearFunction);

    public override Tensor forward(autograd.AutogradContext ctx, params object[] vars)
    {
        var input = (Tensor)vars[0];
        var weight = (Tensor)vars[1];
        var bias = vars.Length == 2 ? null : (Tensor)vars[2];

        ctx.save_for_backward([input, weight, bias]);

        var output = input.mm(weight.t());
        if (bias is not null && !bias.IsInvalid)
            output += bias.unsqueeze(0).expand_as(output);

        return output;
    }
    
        public override List<Tensor> backward(autograd.AutogradContext ctx, List<Tensor> grad_outputs)
    {
        var saved = ctx.get_saved_variables();
        var input = saved[0];
        var weight = saved[1];
        var bias = saved.Count == 2 ? null : saved[2];

        var grad_output = grad_outputs[0];
        var grad_input = grad_output.mm(weight);
        var grad_weight = grad_output.t().mm(input);
        var grad_bias = bias is null || bias.IsInvalid ? null : grad_output.sum(0);

        return [grad_input, grad_weight, grad_bias];
    }
}
```

**Comment by shaltielshmid** [comment]
There also is an inheritable class `torch.autograd.MultiTensorFunction` which looks exactly the same except the forward function returns a `List<Tensor>` instead of `Tensor`.

File: src/TorchSharp/AutogradFunction.cs [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+                        _applyFuncReturnArray?.Dispose();
+                        _applyFuncReturnArray = null;
+                        _outputCache?.ForEach(t => t?.Dispose());
+                        _outputCache = null;
+                    }
+                }
+
+                ~Node()
+                {
+                    Dispose(true);
```
Comment (NiklasGustafsson) [inline src/TorchSharp/AutogradFunction.cs:293]: Why are you passing 'true' from a finalizer?

File: src/TorchSharp/AutogradFunction.cs [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+            /// </summary>
+            public abstract class Function<T> where T : Function<T>, new()
+            {
+                internal static T Instance { get; set; } = new();
+
+                /// <summary>
+                /// When calling the function, the user should call Function.apply and not the forward function, so that
+                /// the computation graph is built correctly.
+                /// </summary>
+                public static List<Tensor> apply_internal(params object[] vars)
```
Comment (NiklasGustafsson) [inline src/TorchSharp/AutogradFunction.cs:428]: What else than Tensor could the parameters be?

File: src/TorchSharp/AutogradFunction.cs [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+                public void set_materialize_grads(bool value) => _materializeGrads = value;
+
+                public bool MaterializeGrads => _materializeGrads;
+
+                /// <summary>
+                /// Get the list of variables that were saved in `forward` using
+                /// `save_for_backward()`. Before returning them to the user, a check is made
+                /// to ensure that they were not modified by any in-place operations.
+                /// </summary>
+                public List<Tensor> get_saved_variables()
```
Comment (NiklasGustafsson) [inline src/TorchSharp/AutogradFunction.cs:360]: When I call 'zero_grad()' on the optimizer I'm using for training, can the saved variables be cleared? Are they?

File: src/TorchSharp/AutogradFunction.cs (hunk 1 above)
Comment (shaltielshmid) [reply src/TorchSharp/AutogradFunction.cs:293]: My bad, will remove and call the regular Dispose().

File: src/TorchSharp/AutogradFunction.cs (hunk 2 above)
Comment (shaltielshmid) [reply src/TorchSharp/AutogradFunction.cs:428]: An int, double, floatt.  
See here an example: https://pytorch.org/cppdocs/api/structtorch_1_1autograd_1_1_function.html

File: src/TorchSharp/AutogradFunction.cs (hunk 1 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/AutogradFunction.cs:293]: No, I think you're supposed to pass 'false'

File: src/TorchSharp/AutogradFunction.cs (hunk 3 above)
Comment (shaltielshmid) [reply src/TorchSharp/AutogradFunction.cs:360]: Yup! When you call zero_grad then the node is disposed, which then in turn disposes the saved variables.

File: src/TorchSharp/AutogradFunction.cs (hunk 1 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/AutogradFunction.cs:293]: https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/dispose-pattern#finalizable-types

File: src/TorchSharp/AutogradFunction.cs (hunk 2 above)
Comment (shaltielshmid) [reply src/TorchSharp/AutogradFunction.cs:428]: Also here: https://pytorch.org/tutorials/advanced/cpp_autograd.html
For example, the MulConstant

File: src/TorchSharp/AutogradFunction.cs (hunk 1 above)
Comment (shaltielshmid) [reply src/TorchSharp/AutogradFunction.cs:293]: Ah, I see. OK, will fix that.

**Comment by NiklasGustafsson** [comment]
All the builds fail with compiler errors:

https://dev.azure.com/dotnet/TorchSharp/_build/results?buildId=102357&view=results

**Comment by NiklasGustafsson** [comment]
Even with your fixes, the Mac C++ compiler doesn't like 'to_optional' in your source code:

```
/Users/runner/work/1/s/src/Native/LibTorchSharp/THSAutograd.cpp(139,20): error G722AAD00: no matching function for call to 'to_optional' [/Users/runner/work/1/s/src/Native/build.proj]
      auto outputs = torch::autograd::to_optional(toTensors<at::Tensor>(outputs_.array, outputs_.size));

/Users/runner/work/1/s/src/Native/LibTorchSharp/Utils.h:41:5: note: expanded from macro 'CATCH'
      x \
      ^
  /Users/runner/work/1/s/bin/obj/AnyCPU.Debug/libtorch-cpu/libtorch-macos-2.1.0cpu/libtorch/include/torch/csrc/autograd/custom_function.h:247:45: note: candidate function not viable: no known conversion from 'std::vector<Tensor>' to 'torch::autograd::Variable &' (aka 'at::Tensor &') for 1st argument
  inline std::vector<c10::optional<Variable>> to_optional(Variable& output) {
                                              ^
  /Users/runner/work/1/s/bin/obj/AnyCPU.Debug/libtorch-cpu/libtorch-macos-2.1.0cpu/libtorch/include/torch/csrc/autograd/custom_function.h:251:45: note: candidate function not viable: expects an lvalue for 1st argument
  inline std::vector<c10::optional<Variable>> to_optional(variable_list& output) {
```

**Comment by shaltielshmid** [comment]
> All the builds fail with compiler errors:
> 
> https://dev.azure.com/dotnet/TorchSharp/_build/results?buildId=102357&view=results

Sorry, my bad. I had used the .NET 8 collection initializer in the tests. (It had compiled and all the tests passed locally by me)
I pushed a fix and confirmed the Azure Pipeline build succeeded (at least on one of them).

**Comment by shaltielshmid** [comment]
> Even with your fixes, the Mac C++ compiler doesn't like 'to_optional' in your source code:
> 
> ```
> /Users/runner/work/1/s/src/Native/LibTorchSharp/THSAutograd.cpp(139,20): error G722AAD00: no matching function for call to 'to_optional' [/Users/runner/work/1/s/src/Native/build.proj]
>       auto outputs = torch::autograd::to_optional(toTensors<at::Tensor>(outputs_.array, outputs_.size));
> 
> /Users/runner/work/1/s/src/Native/LibTorchSharp/Utils.h:41:5: note: expanded from macro 'CATCH'
>       x \
>       ^
>   /Users/runner/work/1/s/bin/obj/AnyCPU.Debug/libtorch-cpu/libtorch-macos-2.1.0cpu/libtorch/include/torch/csrc/autograd/custom_function.h:247:45: note: candidate function not viable: no known conversion from 'std::vector<Tensor>' to 'torch::autograd::Variable &' (aka 'at::Tensor &') for 1st argument
>   inline std::vector<c10::optional<Variable>> to_optional(Variable& output) {
>                                               ^
>   /Users/runner/work/1/s/bin/obj/AnyCPU.Debug/libtorch-cpu/libtorch-macos-2.1.0cpu/libtorch/include/torch/csrc/autograd/custom_function.h:251:45: note: candidate function not viable: expects an lvalue for 1st argument
>   inline std::vector<c10::optional<Variable>> to_optional(variable_list& output) {
> ```

I see the issue. I don't have a mac machine to test on, but I think I see the issue. I pushed an attempt to fix.

**Comment by NiklasGustafsson** [comment]
Looks good now!

File: test/TorchSharpTest/TestAutogradFunction.cs [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+                LossIsClose(loss, 0.4513f);
+            }
+        }
+
+        [Fact]
+        public void TestCustomLinearXORLearn_Parallel_CPU()
+        {
+            Enumerable.Range(0, 20).AsParallel().ForAll(i => {
+                float loss = TrainXOR(torch.CPU);
+                LossIsClose(loss, 0.4513f);
```
Comment (NiklasGustafsson) [inline test/TorchSharpTest/TestAutogradFunction.cs:117]: How did you establish these losses as being the expected ones?

Comment (shaltielshmid) [reply test/TorchSharpTest/TestAutogradFunction.cs:117]: I ran it myself on my machine, watched the loss go down.

File: test/TorchSharpTest/TestAutogradFunction.cs [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+            });
+        }
+
+        [Fact]
+        private void TestCustomLinearFunctionWithGC()
+        {
+            var x = torch.randn(new long[] { 2, 3 }).requires_grad_();
```
Comment (NiklasGustafsson) [inline test/TorchSharpTest/TestAutogradFunction.cs:124]: Why does the input require gradients?

File: test/TorchSharpTest/TestAutogradFunction.cs (hunk 4 above)
Comment (NiklasGustafsson) [reply test/TorchSharpTest/TestAutogradFunction.cs:117]: Does running it in Python come up with a similar result?

Comment (shaltielshmid) [reply test/TorchSharpTest/TestAutogradFunction.cs:117]: Good question. I'll run that tomorrow to confirm.

File: test/TorchSharpTest/TestAutogradFunction.cs (hunk 5 above)
Comment (shaltielshmid) [reply test/TorchSharpTest/TestAutogradFunction.cs:124]: Also good catch. I'll remove that. 
I took the base code from https://pytorch.org/tutorials/advanced/cpp_autograd.html , and the example there included gradient on X, but probably just to illustrate the point . 
I'll remove it.

File: test/TorchSharpTest/TestAutogradFunction.cs (hunk 4 above)
Comment (shaltielshmid) [reply test/TorchSharpTest/TestAutogradFunction.cs:117]: https://gist.github.com/shaltielshmid/a4dd4e1f1f0938dc9dbc17b32432a4f3

Confirmed.

**Comment by shaltielshmid** [comment]
Update: I modified the SingleTensorFunction so that the backward function accepts a single Tensor for grad_output instead of a list of tensors.

**Review by NiklasGustafsson — APPROVED** [review]
